rfcs/20250520-async-threadpool-support/README.md

```diff
@@ -0,0 +1,157 @@
+# Proposal to support asynchronous threadpool runtime
+
+## Background
+
+XLA is switching to a new threapool runtime on CPU. In particular:
+- it has an asynchronous `parallel_for` implementation, meaning the
+  `parallel_for` function call returns while submitted tasks are still
+  running on the threadpool.
+- XLA will have a single threadpool for both primitive creation and
+  execution. This allows to overlap primitive creation time with
+  primitive execution. As a result, multiple primitives will be
+  submitted from a threadpool to that same threadpool, so parallelism
+  will always be nested.
+
+We have some properties that we can leverage on oneDNN side:
+- Primitive and memory object lifetimes will be maintained by thunk
+  runtime while execution happens: no need to reference count oneDNN
+  objects to extend their lifetime.
+- consecutive `parallel_for` calls happen in order (similar to sycl
+  in\_order queues). We can leverage that to implicitly synchronize
+  multiple parallel regions.
+
+## API changes
+
+### Waiting on threadpool tasks completion
+
+Currently the external threadpool interface that users need to
+implement already exposes flags, to inform oneDNN implementation of
+the threadpool implementation properties.  The only flag we expose as
+of now is the `ASYNCHRONOUS` flag, which when set, implies that the
+threadpool iface implementation has a non-blocking implementation of
+the parallel method.
+
+However, until now, when the threadpool iface implementation was
+ASYNCHRONOUS, we used to explicit synchronize inside oneDNN library,
+and we plan on removing that synchronization. With that change, we
+need to add a wait() method to the threadpool iface, so that we can
+properly handle waiting through threadpool in stream:wait() method.
+
+Additions to threadpool_iface.hpp:
+```c++
+struct threadpool_iface {
+     // Does nothing if SYNCHRONOUS, waits for all jobs for ASYNCHRONOUS 
+     virtual void wait() = 0;
```

Comment:
Who is expected to call tp->wait()?

Reply:
we call it internally to implement stream->wait().
This is used to wait for work completion in our tests and examples.